File: go/basics/part0/stubs.go
package main

import (
	"fmt"
	"math/rand"
)

func MaybeCrash(probability float32) error {
	if rand.Float32() < probability { // 50% chance of failure
		fmt.Printf("👻 A failure happened!")
		return fmt.Errorf("a failure happened")
	}
	return nil
}

// CreateSubscription Simulates calling a subscription API, with a random probability of API downtime.
func CreateSubscription(userID, subscription, paymentRef string) error {
	err := MaybeCrash(0.3)
	if err != nil {
		return err
	}
	fmt.Printf(">>> Creating subscription %s for user %s\n", subscription, userID)
	return nil
}

// CreateRecurringPayment Simulates calling a payment API, with a random probability of API downtime.
func CreateRecurringPayment(creditCard string, paymentID interface{}) (string, error) {
	err := MaybeCrash(0.3)
	if err != nil {
		return "", err
	}
	fmt.Printf(">>> Creating recurring payment %v\n", paymentID)
	return "payment-reference", nil
}

File: rust/basics/src/utils/stubs.rs
use rand::random;
use anyhow::{anyhow, Result};
use restate_sdk::errors::HandlerError;
use log::{error, info};
use serde::{Deserialize, Serialize};

#[derive(Deserialize)]
pub struct SubscriptionRequest {
    pub user_id: String,
    pub credit_card: String,
    pub subscriptions: Vec<String>,
}

#[derive(Deserialize, Serialize)]
pub struct User {
    pub name: String,
    pub email: String,
}

#[derive(Deserialize, Serialize)]
pub struct BankAccountCharge {
    pub amount: f64,
    pub account: String,
}

pub async fn create_subscription(user_id: &str, subscription: &str, _payment_ref: &str) -> Result<String, HandlerError> {
    if random::<f32>() < 0.3 {
        error!("👻 Failed to create subscription: {} - {}", user_id, subscription);
        return Err(HandlerError::from(anyhow!("👻 Failed to create subscription: {} - {}", user_id, subscription)));
    }
    info!("Creating subscription {} for user {}", subscription, user_id);
    Ok("SUCCESS".to_string())
}

pub async fn create_recurring_payment(_credit_card: &str, payment_id: &str) -> Result<String, HandlerError> {
    if random::<f32>() < 0.5 {
        error!("👻 Failed to create recurring payment: {} ", payment_id);
        return Err(HandlerError::from(anyhow!("👻 Failed to create recurring payment: {}", payment_id)));
    }
    info!("Creating recurring payment {}", payment_id);
    Ok("payment-reference".to_string())
}

pub async fn create_user_entry(entry: &User) -> Result<(), HandlerError> {
    info!("Creating user entry for {}", entry.name);
    Ok(())
}

pub async fn send_email_with_link(user_id: &str, user: &User, secret: &str) -> Result<(), HandlerError> {
    info!("Sending email to {} with secret {}. \n
    To simulate a user clicking the link, run the following command: \n
    curl localhost:8080/SignupWorkflow/{}/click -H 'content-type: application/json' -d '\"{}\"'",
             user.email, secret, user_id, secret);
    Ok(())
}

pub async fn charge_bank_account(_payment_deduplication_id: &str, _payment: &f32) -> Result<(), HandlerError> {
    Ok(())
}